crates/sui-core/src/db_checkpoint_handler.rs

```diff
@@ -202,90 +212,58 @@ impl DBCheckpointHandler {
             }
             candidate_epoch += 1
         }
-        Ok(candidate_epoch as i64)
+        missing_epochs.push(candidate_epoch);
+        Ok(missing_epochs)
     }
-    async fn upload_db_checkpoint_to_object_store(&self) -> Result<()> {
+    async fn upload_db_checkpoint_to_object_store(&self, missing_epochs: Vec<u32>) -> Result<()> {
+        let last_missing_epoch = missing_epochs.last().cloned().unwrap_or(0);
         let local_checkpoints_by_epoch = self
             .read_checkpoint_dir(self.input_object_store.clone())
             .await?;
-        let remote_checkpoints_by_epoch = self
-            .read_checkpoint_dir(self.output_object_store.clone())
-            .await?;
-
-        let next_epoch = if let Some((last_epoch, path)) =
-            remote_checkpoints_by_epoch.iter().next_back()
-        {
-            let success_marker = path.child(SUCCESS_MARKER);
-            let get_result = self.output_object_store.get(&success_marker).await;
-            match get_result {
-                Ok(_) => last_epoch + 1,
-                Err(Error::NotFound { .. }) => {
-                    // delete this path recursively and try uploading the last epoch db checkpoint again
-                    delete_recursively(
-                        path,
-                        self.output_object_store.clone(),
-                        NonZeroUsize::new(20).unwrap(),
-                    )
-                    .await?;
-                    *last_epoch
-                }
-                Err(err) => {
-                    return Err(anyhow!("Failed to determine if last checkpoint was uploaded successfully with error: {:?}", err));
+        let mut dirs: Vec<_> = local_checkpoints_by_epoch.iter().collect();
+        dirs.sort_by_key(|(epoch_num, _path)| *epoch_num);
```

Comment:
Curious if it would make sense to sort  descending here so that the most recent epoch snapshot is generated first? Or perhaps special case that one? No strong opinions here though

Reply:
I thought of it as well but wasn't sure if the refactor was worth the effort (it'd require changing tests and maybe other corner cases I haven't thought through). In practice it would be useful if we have a lot of old checkpoints locally and remote is empty or has few checkpoints. But that seems highly unlikely. If we start generating snapshots from genesis, uploading it to s3 can easily keep up such that there would be almost no build up locally.